python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,258 @@
+        fields = list(set(fields + ["image", "sg_status_list", "description"]))
+
+        ## folder icon
+        self._default_icon = QtGui.QIcon(QtGui.QPixmap(":/tk-framework-shotgunutils/icon_Folder.png"))
+
+        ShotgunModel.__init__(self, 
+                              parent = parent,
+                              download_thumbs = download_thumbs,
+                              schema_generation = schema_generation,
+                              bg_load_thumbs = bg_load_thumbs,
+                              bg_task_manager = bg_task_manager)
+
+        # load the data from the cache:
+        self._load_data(entity_type, filters, hierarchy, fields)
+
+    def destroy(self):
+        """
+        Call to clean-up the model when it is finished with
+        """
+        ShotgunModel.destroy(self)
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+        self._default_icon = None
+
+    def get_entity_icon(self, entity_type):
+        """
+        Retrieve the icon for the specified entity type if available.
+
+        :param entity_type: The entity type to retrieve the icon for
+        :returns:           A QIcon if an icon was found for the specified entity
+                            type, otherwise None.
+        """
+        icon = None
+        if entity_type in self._entity_icons:
+            # we've previously asked for the icon
+            icon = self._entity_icons[entity_type]
+        else:
+            # see if we have the icon in the resources:
+            icon_path = ":/tk-framework-shotgunutils/icon_%s_dark.png" % entity_type
+            if QtCore.QFile.exists(icon_path): 
+                # create the new icon from this resource:
+                icon = QtGui.QIcon(QtGui.QPixmap(icon_path))
+            self._entity_icons[entity_type] = icon
+
+        return QtGui.QIcon(icon) if icon else None
+
+    def get_entities(self, item):
+        """
+        Get entities for the current item by traversing up the tree and pulling 
+        entity information from each item if possible
+
+        :param item:    The item to find entities for
+        :returns:       A list of entities in the order they were found starting
+                        from the specified item.
+        """
+        entities = []
+        current_item = item
+
+        # first, if this is a leaf item then it will represent an entity:
+        sg_data = current_item.get_sg_data()
+        if sg_data:
+            entities.append(sg_data)
+            current_item = current_item.parent()
+
+        # now walk up the tree and look for an entity in the fields of the 
+        # parent items:
+        while current_item:
+            field_data = get_sanitized_data(current_item, self.SG_ASSOCIATED_FIELD_ROLE)
+            field_value = field_data.get("value")
+            if (field_value 
+                and isinstance(field_value, dict) 
+                and "id" in field_value 
+                and "type" in field_value):
```

Comment:
Just throwing that out there, I'm curious to know what you think: on multi line ifs I prefer putting the closing ): on the next line aligned with the if, this way it's more obvious where the conditional ends and where the indented block starts. That being said, don't bother going through the code to fix this. :)